core/test/multi_index_test.cpp

```diff
@@ -100,22 +143,122 @@ TEST_F(MultiIndexTest, multiple_data_indices) {
 
 TEST_F(MultiIndexTest, advance_multiple_data_indices) {
   MultiIndex index(yx, x, y);
-  index.advance(1);
+  index.set_index(1);
   check(index, {1, 0, 1, 0, 1}, {0, 1, 1, 2, 2});
-  index.advance(1);
+  index.set_index(2);
   check(index, {0, 1, 0, 1}, {1, 1, 2, 2});
 }
 
 TEST_F(MultiIndexTest, advance_slice_middle) {
   MultiIndex index(xz, xyz);
-  index.advance(2);
+  index.set_index(2);
   check(index, {2, 3, 12, 13, 14, 15});
-  index.advance(3);
+  index.set_index(5);
   check(index, {13, 14, 15});
 }
 
 TEST_F(MultiIndexTest, advance_slice_and_broadcast) {
   MultiIndex index(xz, xy);
-  index.advance(2);
+  index.set_index(2);
   check(index, {0, 0, 3, 3, 3, 3});
 }
+
+TEST_F(MultiIndexTest, 1d_array_of_1d_buckets) {
+  const Dim dim = Dim::Row;
+  Dimensions buf{{dim}, {7}}; // 1d cut into two sections
+  // natural order no gaps
+  check_with_buckets(buf, dim, {{0, 3}, {3, 7}}, x, x, {0, 1, 2, 3, 4, 5, 6});
+  // gap between
+  check_with_buckets(buf, dim, {{0, 3}, {4, 7}}, x, x, {0, 1, 2, 4, 5, 6});
+  // gap at start
+  check_with_buckets(buf, dim, {{1, 3}, {3, 7}}, x, x, {1, 2, 3, 4, 5, 6});
+  // out of order
+  check_with_buckets(buf, dim, {{4, 7}, {0, 4}}, x, x, {4, 5, 6, 0, 1, 2, 3});
+}
+
+TEST_F(MultiIndexTest, 1d_array_of_2d_buckets) {
+  Dimensions buf{{Dim("a"), Dim("b")}, {2, 3}}; // 2d cut into two sections
+  // cut along inner
+  check_with_buckets(buf, Dim("b"), {{0, 1}, {1, 3}}, x, x, {0, 3, 1, 2, 4, 5});
```

Comment:
Intuitively, i would have thought that this would give `{0, 3, 1, 4, 2, 5}`:
If you have a layout:
```
0 3
1 4
2 5
```
taking the first row gives the first `0 3` and then 2nd and 3rd rows yield `1 4` and `2 5`?

Reply:
Or maybe because the second bucket is spanning two rows, it always iterates along the outer dim first, so it would read `1 2` and then `4 5`?

Reply:
Yes, it is spanning two rows. As usual, it iterates the *inner* dim (here `b`) first, so numbers come in order: 2 after 1